src/NuGet.Core/NuGet.Protocol/Model/GetVulnerabilityInfoResult.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol.Model
+{
+    /// <summary>Represents the result of <see cref="IVulnerabilityInfoResource.GetVulnerabilityInfoAsync(Core.Types.SourceCacheContext, Common.ILogger, System.Threading.CancellationToken)"/></summary>
+    /// <remarks>Vulnerability data from a package source typically is spread across multiple files. Therefore, it's
+    /// possible for some data to be retrieved successfully, while other data failed. This type allows partial data
+    /// to be returned along with error information.</remarks>
+    public sealed class GetVulnerabilityInfoResult
+    {
+        /// <summary>The known package vulnerabilities.</summary>
+        /// <remarks>The outer IReadOnlyList represents the number of files the package source split the vulnerability data into.
+        /// The IReadOnlyDictionary's key is the package ID.
+        /// The inner IReadOnlyList is all of the vulnerabilities known for the package ID.
+        /// <para>May be null if no vulnerability data was parsed successfully.</para></remarks>
+        public IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? KnownVulnerabilities { get; }
+
+        /// <summary>Any exceptions while reading vulnerability data.</summary>
+        /// <remarks>May be null if no exceptions occurred.</remarks>
+        public AggregateException? Exceptions { get; }
```

Comment:
`AggregateException` contains a list of exceptions. Previously I had the type as `IReadOnlyList<Exception>`, but thought maybe `AggregateException` might be "better" since it's a built-in type for basically the same purpose.

I think that returning the "real" exception when there's only 1, and returning AggregateException when there are multiple is more confusing, because then it ends up like JavaScript/PowerShell "is there one or more than one" code everywhere since you never really know what you're going to get.  An array of 1 allows code to always work the same way, and maybe sometimes it'll only iterate a single item.

Having written all that, since you're asking, clearly it's not obvious.  Should I change it back to `IReadOnlyList<Exception>`?  Is all my justification unconvincing and you think I should still use `Exception` anyway?

Reply:
I don't think it would make sense for this to return a list of `Exception` objects.  I'm only saying it could return `Exception` and the caller could check if its an `AggregateException`.  Otherwise our code has to always wrap exceptions in an `AggregateException`.  I don't see anywhere else in our code where we return an `AggregateException`, only cases where we return `Exception`.
```
src\NuGet.Core\NuGet.PackageManagement\IDE\PackageRestoreFailedEventArgs.cs:        public Exception Exception { get; private set; }
src\NuGet.Core\NuGet.Protocol\Plugins\FaultedPluginEventArgs.cs:        public Exception Exception { get; }
src\NuGet.Core\NuGet.Protocol\Plugins\PluginCreationResult.cs:        public Exception Exception { get; }
src\NuGet.Core\NuGet.Protocol\Plugins\ProtocolErrorEventArgs.cs:        public Exception Exception { get; }
```

In my opinion, it would be best to return a more abstract type instead of a concrete type and is a more standard pattern.  The XML documentation can indicate to callers that it could be an `AggregateException`.

This would mean that `GetVulnerabilityInfoAsync` would need to still keep track of the exceptions but if only one was thrown return that, otherwise return an `AggregateException`.  We could also just keep its implementation and this property is just always an `Exception` but current implementation is really always an `AggregateException`.  Just my vote.

Reply:
> I'm only saying it could return Exception and the caller could check if its an AggregateException.

And this is what I'm referring to that frustrates developers with PowerShell (and maybe also Javascript?)  when a method returns more than 1 items, the return type is an array, but when it returns exactly one item, it's not an array. For example the version numbering problem we had a few weeks ago when MSBuild started adding empty lines to its output, and `output.Trim()` in PowerShell silently trimmed each line individually (since it was treated as an array), rather than trimming the new line characters out of we expected to be a single string.

It sounds easy to "check if it's an AggregateException", but so is "check if the PowerShell method returned an array or a single item". Not everyone will read the docs, and when it behaves differently to the simple tests people tried while developing, bugs will happen.  Good APIs have few surprises, and that was my intention here. It's designed to potentially have more than 1 error, so the data type represents that.

[The BCL's Task.Exception](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.exception?view=net-7.0#system-threading-tasks-task-exception) is a kind of precedent. It has type `AggregateException`, even though a task usually represents a single code path execution, not a `Task.WhenAll` or similar.

> Otherwise our code has to always wrap exceptions in an AggregateException

I don't follow. Why would other code, that have different design goals, need to wrap all exceptions in an AggregateException?

>  I don't see anywhere else in our code where we return an AggregateException, only cases where we return Exception.

And I'm trying to do something different to basically anything else already in the NuGet repo:  graceful degradation when there's a problem. The closest precedent I can think of is when a package version can't be found or downloaded during the first stage of graph resolution. It doesn't instantly crash the whole restore process, it still allows other packages to be restored, but the project will error out since part of the graph couldn't be determined. However, a failed package doesn't have the same "partial data" concept that vulnerability data does (multiple pages). Other examples of graceful error handling are the second stage of graph resolution, or pack validations, but in both of those cases, they're types of validation on existing data, its not simultaneously creating new data types to return, at the same time as detecting errors gracefully.

If there's a problem with one vulnerability page in a server, I still want to use the successful pages to check for known vulnerabilities. If one server's vulnerability data is completely broken, I still want to use different server's known vulnerabilities.

Reply:
Okay I suppose this is fine, I don't feel that strongly about it.  It seems weird but it does fit the implementation where an exception isn't thrown and potentially incomplete results are returned.